### PR TITLE
fix: [Bug] Pipeline parameters (ParameterInteger, ParameterString) fail in ModelTrain (5504)

### PR DESCRIPTION
**Pull request by aviruthen**
## Description

The `safe_serialize` function in `sagemaker-train/src/sagemaker/train/utils.py` already has a `PipelineVariable` isinstance check (lines 185-186), but the user's SDK version (3.3.1) may not have this fix. Additionally, the fallback `except TypeError: return str(data)` block is dangerous because `PipelineVariable.__str__()` intentionally raises `TypeError`, meaning if the isinstance check ever fails (e.g., import path mismatch, reload issues), the except block will re-raise. The fix needs to: (1) ensure the PipelineVariable isinstance check is solid, (2) make the except fallback more robust by catching the case where `str()` also raises TypeError, and (3) add missing unit tests for `safe_serialize` from the train utils module covering PipelineVariable inputs. There are no existing tests for the `safe_serialize` in `sagemaker-train/src/sagemaker/train/utils.py`.

## Related Issue

Related issue: 5504

## Changes Made

- `sagemaker-train/src/sagemaker/train/utils.py`
- `sagemaker-train/tests/unit/train/test_utils.py`
- `sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py`

## AI-Generated PR

This PR was automatically generated by the PySDK Issue Agent.
- Confidence score: 85%
- Classification: bug
- SDK version target: V3

## Merge Checklist

- [x] Changes are backward compatible
- [x] Commit message follows `prefix: description` format
- [x] Unit tests added/updated
- [ ] Integration tests added (if applicable)
- [ ] Documentation updated (if applicable)

### CODE REVIEW

File: sagemaker-train/src/sagemaker/train/utils.py [hunk 1]
```diff
@@ -192,7 +192,16 @@ def safe_serialize(data):
     try:
         return json.dumps(data)
     except TypeError:
-        return str(data)
+        try:
+            return str(data)
+        except TypeError:
+            # PipelineVariable.__str__ raises TypeError by design.
+            # If the isinstance check above didn't catch it (e.g. import
+            # path mismatch), fall back to returning the object directly
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py]: Using `hasattr(data, 'expr')` as a duck-typing check for PipelineVariable is fragile — any object with an `expr` attribute would match this condition. Since there's already a `PipelineVariable` isinstance check earlier in the function (lines 185-186), a more robust approach would be to repeat that isinstance check here, or better yet, import and check against the specific parameter types. This would also be more maintainable.

Alternatively, consider catching `TypeError` more broadly:
```python
except TypeError:
    try:
        return str(data)
    except TypeError:
        # PipelineVariable.__str__ raises TypeError by design.
        # Return the object directly so pipeline serialization can handle it.
        return data
```
Returning `data` unconditionally in the inner except is safer than a duck-type check that could silently pass through unrelated objects. If the object truly can't be serialized or stringified, it's likely a PipelineVariable that should be passed through. If you want to be defensive, re-use the `PipelineVariable` isinstance check instead of `hasattr`.

File: sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py [hunk 2]
```diff
@@ -176,3 +177,61 @@ def test_training_image_rejects_invalid_type(self):
...
+
+class TestModelTrainerHyperparametersPipelineVariable:
+    """Test that PipelineVariable objects in hyperparameters survive safe_serialize."""
+
+    def test_hyperparameters_with_pipeline_variable_integer(self):
+        """ParameterInteger in hyperparameters should be passed through as-is."""
+        max_depth = ParameterInteger(name="MaxDepth", default_value=5)
+        trainer = ModelTrainer(
+            training_image=DEFAULT_IMAGE,
+            role=DEFAULT_ROLE,
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py:189]: The `trainer` variable is created but never used in this test — the test only exercises `safe_serialize` directly. Either remove the unused `ModelTrainer` instantiation (and test `safe_serialize` in `test_utils.py` where it already lives), or actually assert something about `trainer.hyperparameters` to verify end-to-end behavior. The same issue applies to `test_hyperparameters_with_pipeline_variable_string` and `test_hyperparameters_with_mixed_pipeline_and_regular_values`.

File: sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py [hunk 3]
```diff
@@ -176,3 +177,61 @@ def test_training_image_rejects_invalid_type(self):
...
+            compute=DEFAULT_COMPUTE,
+            stopping_condition=DEFAULT_STOPPING,
+            output_data_config=DEFAULT_OUTPUT,
+            hyperparameters={"max_depth": max_depth},
+        )
+        # safe_serialize should return the PipelineVariable object directly
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py]: Imports should be at the top of the file, not inside test functions. Move `from sagemaker.train.utils import safe_serialize` to the module-level imports. This applies to all three new test methods.

File: sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py [hunk 4]
```diff
@@ -176,3 +177,61 @@ def test_training_image_rejects_invalid_type(self):
                 stopping_condition=DEFAULT_STOPPING,
                 output_data_config=DEFAULT_OUTPUT,
             )
+
+
+class TestModelTrainerHyperparametersPipelineVariable:
+    """Test that PipelineVariable objects in hyperparameters survive safe_serialize."""
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py]: These tests in `test_model_trainer_pipeline_variable.py` are essentially duplicates of the tests already added in `test_utils.py`. The `safe_serialize` unit tests in `test_utils.py` are sufficient for testing the serialization behavior. These tests should instead verify the *integration* — that `ModelTrainer` correctly preserves PipelineVariable objects in its hyperparameters dict after construction, rather than just re-testing `safe_serialize`.

File: sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py [hunk 5]
```diff
@@ -26,13 +26,14 @@
 
 from sagemaker.core.helper.session_helper import Session
 from sagemaker.core.helper.pipeline_variable import PipelineVariable, StrPipeVar
-from sagemaker.core.workflow.parameters import ParameterString
+from sagemaker.core.workflow.parameters import ParameterString, ParameterInteger, ParameterFloat
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py]: `ParameterFloat` and `PipelineSession` are imported but `ParameterFloat` is never used in the new tests, and `PipelineSession` is also unused. Remove unused imports.

File: sagemaker-train/tests/unit/train/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+# ---------------------------------------------------------------------------
+# Regular / primitive inputs
+# ---------------------------------------------------------------------------
+
+def test_safe_serialize_with_string_returns_string_as_is():
+    assert safe_serialize("hello") == "hello"
+    assert safe_serialize("12345") == "12345"
+
+
+def test_safe_serialize_with_int_returns_json_string():
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_utils.py:63]: The assertion `safe_serialize("hello") == "hello"` relies on the early-return `isinstance(data, str)` check in `safe_serialize`. This is fine, but consider also adding a test for a string that looks like JSON (e.g., `'{"key": "value"}'`) to ensure it's returned as-is and not double-serialized.

File: sagemaker-train/src/sagemaker/train/utils.py [hunk 7]
```diff
@@ -192,7 +192,16 @@ def safe_serialize(data):
     try:
         return json.dumps(data)
     except TypeError:
-        return str(data)
+        try:
+            return str(data)
+        except TypeError:
+            # PipelineVariable.__str__ raises TypeError by design.
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py:202]: Returning `data` (the raw PipelineVariable object) from a function that otherwise returns `str` creates an inconsistent return type. The function signature/contract implies it returns a string, but now it can return a `PipelineVariable` object. This should be documented in the function's docstring and ideally reflected in the type annotation (e.g., `-> str | PipelineVariable`). Callers need to know they might get a non-string back.

File: sagemaker-train/src/sagemaker/train/utils.py [hunk 8]
```diff
@@ -168,22 +168,26 @@ def convert_unassigned_to_none(instance) -> Any:
     return instance
 
 
-def safe_serialize(data):
+def safe_serialize(data) -> "str | PipelineVariable":
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py:171]: The return type `"str | PipelineVariable"` is a forward reference string, but this module should use `from __future__ import annotations` at the top to enable PEP 604 union syntax properly. Also, this is a behavioral change to the function's contract — previously it always returned `str`, now it can return `PipelineVariable`. This could cause `TypeError` in downstream callers that expect a `str` (e.g., calling `.encode()`, concatenation, etc.). Have you audited all call sites of `safe_serialize` to ensure they handle a `PipelineVariable` return value correctly?

Suggestion: Add `from __future__ import annotations` at the module top, and change the annotation to:
```python
def safe_serialize(data) -> str | PipelineVariable:
```
Also consider using `Union[str, PipelineVariable]` with an explicit import if `from __future__ import annotations` is not already used in this module.

File: sagemaker-train/src/sagemaker/train/utils.py [hunk 9]
```diff
@@ -192,7 +196,14 @@ def safe_serialize(data):
     try:
         return json.dumps(data)
     except TypeError:
-        return str(data)
+        try:
```
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py:199]: The nested try/except is a reasonable defensive measure, but catching a bare `TypeError` from `str(data)` and silently returning the raw object is risky. This means *any* object whose `__str__` raises `TypeError` (not just `PipelineVariable`) will be returned as-is, potentially causing unexpected behavior downstream. Consider being more explicit:

```python
except TypeError:
    try:
        return str(data)
    except TypeError:
        if isinstance(data, PipelineVariable):
            return data
        raise
```

This way, only known `PipelineVariable` objects get the pass-through treatment, and truly broken objects still raise.

File: sagemaker-train/src/sagemaker/train/utils.py (hunk 8 above)
Comment (mufaddal-rohawala) [inline sagemaker-train/src/sagemaker/train/utils.py:171]: Missing type annotation for the `data` parameter. Per SDK conventions, all public functions must have type annotations for parameters and return types.

```python
def safe_serialize(data: Any) -> str | PipelineVariable:
```

File: sagemaker-train/tests/unit/train/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Unit tests for sagemaker.train.utils – specifically safe_serialize."""
+from __future__ import absolute_import
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_utils.py:14]: Per PEP 484 / SDK conventions, new modules should use `from __future__ import annotations` instead of `from __future__ import absolute_import` (which is a Python 2 artifact and unnecessary in Python 3).

File: sagemaker-train/tests/unit/train/test_utils.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+def test_safe_serialize_with_none_returns_json_null():
+    assert safe_serialize(None) == "null"
+
+
+def test_safe_serialize_with_list_returns_json_string():
+    assert safe_serialize([1, 2, 3]) == "[1, 2, 3]"
+
+
+def test_safe_serialize_with_empty_string():
+    assert safe_serialize("") == ""
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_utils.py:96]: Good test coverage overall! Consider adding a test for the specific edge case this PR is fixing — where `isinstance` check for `PipelineVariable` might fail (e.g., import path mismatch). You could mock `isinstance` or create a mock object whose `__str__` raises `TypeError` to verify the nested except path:

```python
def test_safe_serialize_with_object_whose_str_raises_typeerror():
    """Objects whose __str__ raises TypeError should be returned as-is."""
    class BadStr:
        def __str__(self):
            raise TypeError("cannot convert")
    obj = BadStr()
    result = safe_serialize(obj)
    assert result is obj
```

This directly tests the new fallback code path added in this PR.

File: sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py [hunk 12]
```diff
@@ -176,3 +176,54 @@ def test_training_image_rejects_invalid_type(self):
...
+            compute=DEFAULT_COMPUTE,
+            stopping_condition=DEFAULT_STOPPING,
+            output_data_config=DEFAULT_OUTPUT,
+            hyperparameters={"max_depth": max_depth},
+        )
+        assert trainer.hyperparameters["max_depth"] is max_depth
+
+    def test_hyperparameters_preserves_pipeline_variable_string(self):
+        """ParameterString in hyperparameters should be preserved in ModelTrainer."""
+        optimizer = ParameterString(name="Optimizer", default_value="sgd")
```
Comment (mufaddal-rohawala) [inline sagemaker-train/tests/unit/train/test_model_trainer_pipeline_variable.py:199]: These tests verify that `ModelTrainer` preserves PipelineVariable objects in the `hyperparameters` dict, but they don't test the actual serialization path (i.e., when `safe_serialize` is called during job creation). Consider adding a test that mocks the training job creation to verify that PipelineVariable values survive the full serialization pipeline, not just assignment.
